### DISCUSSION TITLE
get a list of roles with their users and permissions

### DISCUSSION THREAD

**Discussion by moecods** [spatie/laravel-permission#1938] [Q&A] [answered]
How can I get a list of roles by their users?

for example i want:

```
role-1:
     permissions:  
                    permission-1
                    permission-2
      users:
           user-1
           user-2
role-2
          .....
```

but i can't get role's users

my code

`$roles = Role::with(['permissions'])->get()->toArray();`

i dont know how add users!

**Comment by erikn69** [comment, marked as answer]
Read sources too always
`$roles = Role::with('permissions','users')->get();`
https://github.com/spatie/laravel-permission/blob/08cc9bdbf89640cb1d64fb01eb9513c1cbb2549d/src/Models/Role.php#L72-L81

**Reply by moecods** [reply]
i used it but i have error:
Error    Class name must be a valid object or a string

Currently, im using below code but i think is not good

```php
$roles = Role::with(['permissions'])->get();
foreach ($roles as $role) {
    if($role->users->isNotEmpty())
        $roles->find($role)->push($role->users);
}
```

**Reply by erikn69** [reply]
If you have only one `guard` you can override relation on roles 
```php
public function users(): BelongsToMany 
 { 
     return $this->morphedByMany( 
         User::class, 
         'model', 
         config('permission.table_names.model_has_roles'), 
         PermissionRegistrar::$pivotRole, 
         config('permission.column_names.model_morph_key') 
     ); 
 } 
```
Or maybe
```php
$roles = Role::with('permissions')->get()->each(function($role){
    $role->load('users');
});
```

**Reply by erikn69** [reply]
> Error Class name must be a valid object or a string

Not a bug
```php
Role::with('permissions','users')->get();
array:15 [▼
  0 => array:10 [▼
    "id" => 1
    "name" => "Admin"
    "guard_name" => "web"
    "permissions" => array:366 [▶]
    "users" => array:9 [▶]
```
Maybe you overwrite construct on Role
https://github.com/spatie/laravel-permission/blob/08cc9bdbf89640cb1d64fb01eb9513c1cbb2549d/src/Models/Role.php#L23-L25
